pinot-core/src/main/java/com/linkedin/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -33,318 +36,266 @@
-      Map<ServerInstance, DataTable> instanceResponseMap) {
-
-    List<Map<String, Serializable>> reducedGroupByResults =
-        aggregationGroupByOperatorService.reduceGroupByOperators(instanceResponseMap);
+  private void attachAggregationResults(@Nonnull BrokerResponseNative brokerResponseNative,
+      @Nonnull List<AggregationInfo> aggregationsInfo, @Nonnull Map<ServerInstance, DataTable> dataTableMap) {
+    int numAggregations = aggregationsInfo.size();
+    List<List<Serializable>> shuffledAggregationResults = shuffleAggregationResults(aggregationsInfo, dataTableMap);
+    List<AggregationResult> reducedAggregationResults = new ArrayList<>(numAggregations);
+    List<AggregationFunction> aggregationFunctions =
+        AggregationFunctionFactory.getAggregationFunction(aggregationsInfo);
+
+    for (int i = 0; i < numAggregations; i++) {
+      String aggregationFunction = aggregationFunctions.get(i).getFunctionName();
+      String formattedValue = AggregationGroupByOperatorService.formatValue(
+          aggregationFunctions.get(i).reduce(shuffledAggregationResults.get(i)));
+      AggregationResult aggregationResult = new AggregationResult(aggregationFunction, formattedValue);
+      reducedAggregationResults.add(aggregationResult);
+    }
 
-    return aggregationGroupByOperatorService.renderAggregationGroupByResult(reducedGroupByResults);
+    brokerResponseNative.setAggregationResults(reducedAggregationResults);
   }
 
   /**
-   * Reduce the aggregationGroupBy response from various servers, and return a list of
-   * AggregationResult objects, that is used to build the BrokerResponseNative object.
+   * Shuffle aggregation results, gather all results for each aggregation function together.
    *
-   * @param brokerRequest
-   * @param aggregationResultsList
-   * @return
+   * @param aggregationsInfo list of aggregation information.
+   * @param dataTableMap map from server to data table.
+   * @return shuffled aggregation results.
    */
-  private List<AggregationResult> reduceOnAggregationResults(BrokerRequest brokerRequest,
-      List<List<Serializable>> aggregationResultsList) {
-    List<AggregationResult> aggregationResults = new ArrayList<AggregationResult>();
-    List<AggregationFunction> aggregationFunctions = AggregationFunctionFactory.getAggregationFunction(brokerRequest);
-
-    for (int i = 0; i < aggregationFunctions.size(); ++i) {
-      String function = aggregationFunctions.get(i).getFunctionName();
-      Serializable value = formatValue(aggregationFunctions.get(i).reduce(aggregationResultsList.get(i)));
+  private List<List<Serializable>> shuffleAggregationResults(@Nonnull List<AggregationInfo> aggregationsInfo,
+      @Nonnull Map<ServerInstance, DataTable> dataTableMap) {
+    int numAggregations = aggregationsInfo.size();
+    int numDataTables = dataTableMap.size();
+    List<List<Serializable>> shuffledAggregationResults = new ArrayList<>(numAggregations);
```

Comment:
This is shuffled, gather all results for the same aggregation together. We do shuffle first, then reduce.